### PR TITLE
executor: Optimize slow log parsing's splitByColon function

### LINKED ISSUE #54538: Slow log parse performance
## Enhancement
Do cpu profile for query like:
```
SELECT
  Digest,
  Query,
  Conn_ID,
  (UNIX_TIMESTAMP(Time) + 0E0) AS timestamp,
  Query_time,
  Mem_max,
  Process_keys
FROM
  `INFORMATION_SCHEMA`.`CLUSTER_SLOW_QUERY`
WHERE
  Time BETWEEN FROM_UNIXTIME(?)
  AND FROM_UNIXTIME(?)
ORDER BY
  Query_time DESC
LIMIT 100;
```
Find the splitByColon takes about 80% cpu times:
img_v3_02cl_7c4b6024-a3b3-4168-b5c4-86858cb0605g

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #54538

Problem Summary:

### What changed and how does it work?
Replace regexp matching with simple string comparison operations. Besides, previously, ":=" inside "{}" are not handled correctly, fix it in this PR also.
In local mannual test, performance will improve about 10x, from 24s to 2.2s for the following sql:
```SELECT   Digest,   Query,   Conn_ID,   (UNIX_TIMESTAMP(Time) + 0E0) AS timestamp,   Query_time,   Mem_max,   Process_keys FROM   `INFORMATION_SCHEMA`.`CLUSTER_SLOW_QUERY` WHERE   Time BETWEEN FROM_UNIXTIME(1720471890)   AND FROM_UNIXTIME(1720515091) ORDER BY   Query_time DESC LIMIT   100;```
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/slow_query.go [hunk 1]
```diff
@@ -523,32 +522,94 @@ func getLineIndex(offset offset, index int) int {
...
+func isLetterOrNumeric(b byte) bool {
+	return ('A' <= b && b <= 'Z') || ('a' <= b && b <= 'z') || ('0' <= b && b <= '9')
+}
 
 // splitByColon split a line like "field: value field: value..."
+// Note:
+// 1. Both field and value string contain only ANSI characters
+// 2. value string may be surrounded by brackets, allowed brackets includes "[]" and "{}",  like {key: value,{key: value}}
+// "[]" can only be nested inside "[]"; "{}" can only be nested inside "{}"
+// 3. value string can't contain ' ' character unless it is inside brackets
```
Comment (xzhangxian1008) [inline pkg/executor/slow_query.go:568]: Is origin implementation also contains these restriction?

Comment (yibin87) [reply pkg/executor/slow_query.go:568]: No, while current slow log satisfy these restrictions and it doesn't seem to be changed frequently in future. Previous implementation provides a broader functionality.

File: pkg/executor/slow_query_test.go [hunk 2]
```diff
@@ -550,6 +550,26 @@ func TestSplitbyColon(t *testing.T) {
...
+		{
+			"1a: {2b,3c} 4d: 5e",
+			[]string{"1a", "4d"},
+			[]string{"{2b,3c}", "5e"},
+		},
+		{
+			"1a: {2b,{3c: 3cc}} 4d: 5e",
+			[]string{"1a", "4d"},
+			[]string{"{2b,{3c: 3cc}}", "5e"},
+		},
```
Comment (xzhangxian1008) [inline pkg/executor/slow_query_test.go:572]: - Add some invalid cases?
- Maybe we can add some cases that multi left brackets stay together such as "{{{" or "[[["

File: pkg/executor/slow_query.go [hunk 3]
```diff
@@ -523,32 +522,94 @@ func getLineIndex(offset offset, index int) int {
...
+				return -1
+			}
+		} else {
+			current++
+		}
+	}
+	return -1
+}
+
+func isLetterOrNumeric(b byte) bool {
```
Comment (xzhangxian1008) [inline pkg/executor/slow_query.go:559]: Maybe we can put this function in pkg/util/stringutil file.

File: pkg/executor/slow_query.go [hunk 4]
```diff
@@ -523,32 +522,94 @@ func getLineIndex(offset offset, index int) int {
...
+	return ('A' <= b && b <= 'Z') || ('a' <= b && b <= 'z') || ('0' <= b && b <= '9')
+}
 
 // splitByColon split a line like "field: value field: value..."
+// Note:
+// 1. Both field and value string contain only ANSI characters
```
Comment (xhebox) [inline pkg/executor/slow_query.go:565]: That is too restricted. E.g. tidb_redact_log may output non-ansi characters later sometime.
I am thinking if we could just use `strings.Index(a, b)` to mimic the original `regexp`.

Comment (yibin87) [reply pkg/executor/slow_query.go:565]: I see, this rule can be removed for current algorithm, I'll change it.
BTW, splitByColon is not the root entry for slow log parsing, for logs that may contain non-ansi characters we can handle it separately if we doneed it.

File: pkg/executor/slow_query_test.go (hunk 2 above)
Comment (yibin87) [reply pkg/executor/slow_query_test.go:572]: Good suggestion, will add it.

File: pkg/executor/slow_query.go (hunk 4 above)
Comment (yibin87) [reply pkg/executor/slow_query.go]: Checked that whitespace character can't be a part of a valid non-whitespace utf8 character. Thus value string can contain non ascii characters except ' '.  And whitespace can be contained inside "{}"/"[]" in value string. For field string, first character should be letters or digits.

File: pkg/executor/slow_query_test.go (hunk 2 above)
Comment (yibin87) [reply pkg/executor/slow_query_test.go:572]: Done

File: pkg/executor/slow_query.go (hunk 3 above)
Comment (yibin87) [reply pkg/executor/slow_query.go:563]: Just leave it in the local file, since it may be changed for slow log parsing only.

**Review by xhebox — APPROVED** [review]

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

**Pull request merged**